### PR TITLE
Compressor: Add UADK document

### PR DESCRIPTION
**Pull request by kevinzs2048**
The #58336 to support UADK in Ceph compressor zlib is merged, this PR is to submit a document for for the UADK and Ceph integration work.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: doc/radosgw/uadk-accel.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
+===============================================
+UADK Acceleration for Compression
+===============================================
+
+UADK is a framework for user application to access hardware accelerator in a
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:5]: UADK is a framework for applications to access hardware accelerators in a

File: doc/radosgw/uadk-accel.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+unified, secure, and efficient way. UADK is comprised of UACCE, libwd and many
+other algorithm libraries for different applications.
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:7]: s/for different applications//

File: doc/radosgw/uadk-accel.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+See `Compressor UADK Support`_.
+
+
+UADK in the Software Stack
+=========================
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:13]: Add an =

File: doc/radosgw/uadk-accel.rst [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+UADK is a general-purpose user space accelerator framework that uses shared
+virtual addressing (SVA) to provide a unified programming interface for hardware
+acceleration of cryptographic and compression algorithms.
+
+UADK includes Unified/User-space-access-intended Accelerator Framework (UACCE),
+which enables hardware accelerators from different vendors that support SVA to
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:20]: s/from different vendors//

File: doc/radosgw/uadk-accel.rst [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+A user can access the hardware accelerators by performing user-mode operations on
+the character devices, or the use of UADK can be done via frameworks that have
+been enabled by others including UADK support (for example, OpenSSL* libcrypto*,
+DPDK, and the Linux* Kernel Crypto Framework).
+
+See `OpenSSL UADK Engine`_.
+
+UADK Environment Setup
+=====================
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:36]: add =

File: doc/radosgw/uadk-accel.rst [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+UADK consists of UACCE, vendors’ drivers, and an algorithm layer. UADK requires the
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:37]: vendor drivers

File: doc/radosgw/uadk-accel.rst [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+hardware accelerator to support SVA, and the operating system to support IOMMU and
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:38]: So if one currently disables IOMMU on EPYCs for better performance, one would want to try it both ways to see which is faster?

File: doc/radosgw/uadk-accel.rst [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+SVA. Hardware accelerators from different vendors are registered as different character
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:39]: Hardware accelerators are registered as character devices with UACCE by kernel-mode drivers.

File: doc/radosgw/uadk-accel.rst [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+             +----------------------+
+             |   HW Accelerators    |
+             +----------------------+
+
+Configuration
+=============
+
+#. Kernel Requirement
+
+User needs to make sure that UACCE is already supported in Linux kernel. The kernel version
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:81]: Users nee to ensure that UACCE is supported by the Linux kernel release in use, which should be 5.9 or later with SVA (Shared Virtual Addressing) enabled.

File: doc/radosgw/uadk-accel.rst [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+should be at least v5.9 with SVA (Shared Virtual Addressing) enabled.
+
+UACCE could be built as module or built-in. Here’s an example to enable UACCE with hardware
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:84]: UACCE may be built as a module or built into the kernel.  Here's an example to build UACCE with hardware accelerators for the HiSilicon Kunpeng platform.

File: doc/radosgw/uadk-accel.rst [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+       CONFIG_ARM_SMMU_V3_SVA=y
+       CONFIG_PCI_PASID=y
+       CONFIG_UACCE=y
+       CONFIG_CRYPTO_DEV_HISI_QM=y
+       CONFIG_CRYPTO_DEV_HISI_ZIP=y
+
+Make sure all these above kernel configurations are selected.
+
+#. UADK enablement
+If the architecture is Aarch64, it will automatically download the UADK source code to build
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:101]: aarch64

File: doc/radosgw/uadk-accel.rst [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+the static library. If on architecture, user can enable it with build parameters `-DWITH_UADK=true`
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:102]: if on which architecture?

File: doc/radosgw/uadk-accel.rst [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+      git clone https://github.com/Linaro/uadk.git
+      cd uadk
+      mkdir build
+      ./autogen.sh
+      ./configure --prefix=$PWD/build
+      make
+      make install
+
+   .. note:: Without –prefix, UADK will be installed to /usr/local/lib by default.
+   If get error:”cannot find -lnuma”, please install the libnuma-dev
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:119]: space after the :
please install the `libnuma-dev` package.

File: doc/radosgw/uadk-accel.rst [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+#. Configure
+
+   Edit the Ceph configuration file (usually ``ceph.conf``) to enable UADK
+   support for *zlib* compression::
+
+         uadk_compressor_enabled=true
+
+   The default value in `global.yaml.in` for uadk_compressor_enabled is false.
```
Comment (anthonyeleven) [inline doc/radosgw/uadk-accel.rst:128]: backticks around false

**Review by anthonyeleven — COMMENTED** [review]
doc requests

**Comment by zdover23** [comment]
> doc requests

I'll correct all these grammar-and-RST-formatting problems. That way, Kevin can focus on your technical questions.

**Comment by Svelar** [comment]
> I'll correct all these grammar-and-RST-formatting problems. That way, Kevin can focus on your technical questions.

Hi @zdover23 , sorry for the pin. Any schedule for this fix? No rush~

**Comment by zdover23** [comment]
> > I'll correct all these grammar-and-RST-formatting problems. That way, Kevin can focus on your technical questions.
> 
> Hi @zdover23 , sorry for the pin. Any schedule for this fix? No rush~

@Svelar, I'll get to this later today.

**Review by zdover23 — APPROVED** [review]

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/60382 - Squid backport
https://github.com/ceph/ceph/pull/60383 - Reef backport
https://github.com/ceph/ceph/pull/60384 - Quincy backport

File: doc/radosgw/uadk-accel.rst (hunk 2 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 3 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 4 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 6 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst:36]: I've made this change.

File: doc/radosgw/uadk-accel.rst (hunk 8 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst:38]: I have made this change.

File: doc/radosgw/uadk-accel.rst (hunk 9 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst:80]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 10 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 11 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This is already present upon my 04 Dec 2024 review.

File: doc/radosgw/uadk-accel.rst (hunk 12 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: This question remains open. @kevinzs2048, @Svelar, what's the answer to the question?

File: doc/radosgw/uadk-accel.rst (hunk 13 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: I've made this change and cleaned up the English in this note.

File: doc/radosgw/uadk-accel.rst (hunk 14 above)
Comment (zdover23) [reply doc/radosgw/uadk-accel.rst]: I've made this change.

**Comment by zdover23** [comment]
Follow-up PR (at long last): https://github.com/ceph/ceph/pull/60953

File: doc/radosgw/uadk-accel.rst (hunk 7 above)
Comment (Svelar) [reply doc/radosgw/uadk-accel.rst:37]: More focus on offloading CPU usage now. Will try later.

File: doc/radosgw/uadk-accel.rst (hunk 12 above)
Comment (Svelar) [reply doc/radosgw/uadk-accel.rst]: on other architectures

**Comment by zdover23** [comment]
@kevinzs2048, does this get backported to Squid, Reef, and Quincy, or to only one or two of those release branches?

**Pull request merged**